stumpy/stump_tiles.py

```diff
@@ -0,0 +1,662 @@
+        sliding window
+
+    cov_a : ndarray
+        The first covariance term relating T_A[i + k + m - 1] and M_T_m_1[i + k]
+
+    cov_b : ndarray
+        The second covariance term relating T_B[i + m - 1] and μ_Q_m_1[i]
+
+    cov_c : ndarray
+        The third covariance term relating T_A[i + k - 1] and M_T_m_1[i + k]
+
+    cov_d : ndarray
+        The fourth covariance term relating T_B[i - 1] and μ_Q_m_1[i]
+
+    μ_Q_m_1 : ndarray
+        Mean of the query sequence, `Q`, relative to the current sliding window and
+        using a window size of `m-1`
+
+    T_A_subseq_isfinite : ndarray
+        A boolean array that indicates whether a subsequence in `T_A` contains a
+        `np.nan`/`np.inf` value (False)
+
+    T_B_subseq_isfinite : ndarray
+        A boolean array that indicates whether a subsequence in `T_B` contains a
+        `np.nan`/`np.inf` value (False)
+
+    T_A_subseq_isconstant : ndarray
+        A boolean array that indicates whether a subsequence in `T_A` is constant (True)
+
+    T_B_subseq_isconstant : ndarray
+        A boolean array that indicates whether a subsequence in `T_B` is constant (True)
+
+    tiles : ndarray
+        An array of two column arrays representing each tile
+
+    tiles_ranges : ndarray
+        A two-dimensional array representing tile indices for each thread
+
+    tile_size : int
+        Maximum length of each tile
+
+    diags_indices : ndarray
+        All diagonal indices within each tile
+
+    thread_idx : int
+        The thread index
+
+    ρ : ndarray
+        The Pearson correlations
+
+    I : ndarray
+        The matrix profile indices
+
+    ignore_trivial : bool
+        Set to `True` if this is a self-join. Otherwise, for AB-join, set this to
+        `False`. Default is `True`.
+
+    Returns
+    -------
+    None
+
+    Notes
+    -----
+    `DOI: 10.1007/s10115-017-1138-x \
+    <https://www.cs.ucr.edu/~eamonn/ten_quadrillion.pdf>`__
+
+    See Section 4.5
+
+    The above reference outlines a general approach for traversing the distance
+    matrix in a diagonal fashion rather than in a row-wise fashion.
+
+    `DOI: 10.1145/3357223.3362721 \
+    <https://www.cs.ucr.edu/~eamonn/public/GPU_Matrix_profile_VLDB_30DraftOnly.pdf>`__
+
+    See Section 3.1 and Section 3.3
+
+    The above reference outlines the use of the Pearson correlation via Welford's
+    centered sum-of-products along each diagonal of the distance matrix in place of the
+    sliding window dot product found in the original STOMP method.
+    """
+    m_inverse = 1.0 / m
+    constant = (m - 1) * m_inverse * m_inverse  # (m - 1)/(m * m)
+    chunk_size = config.STUMPY_DIAGONALS_CHUNK_SIZE
+
+    iter_ranges = np.zeros((chunk_size, 3), dtype=np.uint64)
+    covariances = np.zeros(chunk_size, dtype=np.float64)
+    diags_indices = np.arange(1 - tile_size, tile_size)
+
+    for tile_idx in range(tiles_ranges[thread_idx][0], tiles_ranges[thread_idx][1]):
```

Comment:
All of this indexing is really hard to read and comprehend what is going on. Can you find some time to clean up/refactor the logic like I did with `_get_tiles`. I'm finding everything hard to follow and I am guessing that there may be an easier way to reference things. We may want to redesign things to be more modular. Right now, it feels like this function is doing too much. At the end of the day, given a tile, the operations on that tile should be the same as any other tile. The function that computes the tile shouldn't even need to know that other tiles exist. You give the function the essential information and the parent function will take the results and put the results back in the right place.

Reply:
@seanlaw to be honest with you I don't know what I was smoking when I wrote this, there's a lot going on here. I think if we're able to talk over a discord call sometime we could best discuss this, and possibly quickly brainstorm the better way to do this. It's really hard to explain here 😵‍💫

Reply:
Yes, I'd be happy to do that. Please don't get me wrong, I think you've done a lot of good work here and demonstrated that there is certainly speed to be gained. We just need to spend some time cleaning it up and then I think we'll be able to spot even more speed gains/optimizations.